README.md
# money

Class for math with money

![example workflow](https://github.com/fregate/money/actions/workflows/flutter.yml/badge.svg)

## Getting Started

Add class in project and use when your project need to work with money amounts.

Create, compare, do math, get as double and string


CHANGELOG.md
# 0.9.0 (2021/10/09)

Initial release
* Creation
* Compare
* Math (add, subtract, multiply, divide)
* Get as String (MoneyFormatter), double (4 decimal places precision) and raw (eg serialization)


test/money_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:money/formatter.dart';

import 'package:money/money.dart';

void main() {
  group('Create and Parse', () {
    test('Const Zero value', () {
      final m = Money.zero;
      expect(m.value, isZero);
      expect(m.raw, isZero);
      expect(m.toString(), '0.00');
    });

    group('Int', () {
      test('Positive', () {
        final m = Money(1);
        expect(m.value, 1.0);
        expect(m.raw, 10000);
        expect(m.toString(), '1.00');
      });
      test('Negative value', () {
        final m = Money(-1);
        expect(m.value, -1.0);
        expect(m.raw, -10000);
        expect(m.toString(), '-1.00');
      });
    });

    group('Double', () {
      test('Positive value', () {
        final m = Money(1.0);
        expect(m.value, 1.0);
        expect(m.raw, 10000);
        expect(m.toString(), '1.00');
      });

      test('Negative value', () {
        final m = Money(-1.0);
        expect(m.value, -1.0);
        expect(m.raw, -10000);
        expect(m.toString(), '-1.00');
      });

      test('Infinity value', () {
        expect(() => Money(double.infinity), throwsA(isA<Error>()));
        expect(
            () => Money(double.negativeInfinity), throwsA(isA<Error>()));
      });

      test('NaN value', () {
        expect(() => Money(double.nan), throwsA(isA<Error>()));
      });
    });

    group('String', () {
      test('Value int', () {
        final m = Money('1');
        expect(m.value, 1.0);
        expect(m.raw, 10000);
        expect(m.toString(), '1.00');
      });

      test('Value double', () {
        final m = Money('1.0');
        expect(m.value, 1.0);
        expect(m.raw, 10000);
        expect(m.toString(), '1.00');
      });

      test('Negative value', () {
        final m = Money('-1.0');
        expect(m.value, -1.0);
        expect(m.raw, -10000);
        expect(m.toString(), '-1.00');
      });

      test('Invalid string value 1', () {
        expect(() => Money('a'), throwsA(isA<FormatException>()));
      });

      test('Invalid string value 2', () {
        expect(() => Money('1.1.1'), throwsA(isA<FormatException>()));
      });

      test('Invalid string value 3', () {
        expect(() => Money('1.-1'), throwsA(isA<FormatException>()));
      });

      test('Invalid string value 4', () {
        expect(() => Money('1.a'), throwsA(isA<FormatException>()));
      });

      test('Double infinity', () {
        expect(() => Money(double.infinity.toString()),
            throwsA(isA<FormatException>()));
      });

      test('Double NaN', () {
        expect(() => Money(double.nan.toString()),
            throwsA(isA<FormatException>()));
      });
    });

    group('Object', () {
      test('Other money', () {
        final m = Money('1.0');
        final Money m2 = m;
        expect(m.value, 1.0);
        expect(m.raw, 10000);
        expect(m2.value, 1.0);
        expect(m2.raw, 10000);
      });

      test('Other money parse', () {
        final m = Money('1.0');
        final m2 = Money(m);
        expect(m.value, 1.0);
        expect(m.raw, 10000);
        expect(m2.value, 1.0);
        expect(m2.raw, 10000);
      });

      test('Invalid object', () {
        expect(
            () => Money(DateTime.now()), throwsA(isA<FormatException>()));
      });
    });

    group('TryParse', () {
      test('Int', () {
        expect(Money.tryParse(1), isNotNull);
      });
      test('Double', () {
        expect(Money.tryParse(1.0), isNotNull);
      });
      test('Money', () {
        expect(Money.tryParse(Money.zero), isNotNull);
      });
      test('String', () {
        expect(Money.tryParse('1.0'), isNotNull);
      });
      test('Invalid double', () {
        expect(Money.tryParse(double.nan), isNull);
        expect(Money.tryParse(double.infinity), isNull);
        expect(Money.tryParse(double.negativeInfinity), isNull);
      });
      test('Invalid string', () {
        expect(Money.tryParse('1.a'), isNull);
      });
      test('Invalid object', () {
        expect(Money.tryParse(DateTime.now()), isNull);
      });
    });
  });

  group('Compare', () {
    group('Int', () {
      test('Operator greater than', () {
        final m = Money(100);
        expect(m > Money.zero, true);
        expect(m > 10, true);
        expect(m > 100, false);
        expect(m > 1000, false);
      });

      test('Operator greater or eq than', () {
        final m = Money(100);
        expect(m >= Money.zero, true);
        expect(m >= 10, true);
        expect(m >= 100, true);
        expect(m >= 1000, false);
      });

      test('Operator less than', () {
        final m = Money(100);
        expect(m < 1000, true);
        expect(m < 100, false);
        expect(m < 10, false);
        expect(m < Money.zero, false);
      });

      test('Operator less or eq than', () {
        final m = Money(100);
        expect(m <= 1000, true);
        expect(m <= 100, true);
        expect(m <= 10, false);
        expect(m <= Money.zero, false);
      });

      test('Operator eq', () {
        final m = Money(100);
        expect(m == 1000, false);
        expect(m == 100, true);
        expect(m == 10, false);
        expect(m == Money.zero, false);
      });

      test('Operator not eq', () {
        final m = Money(100);
        expect(m != 1000, true);
        expect(m != 100, false);
        expect(m != 10, true);
        expect(m != Money.zero, true);
      });
    });

    group('Double', () {
      test('Operator greater than', () {
        final m = Money(100);
        expect(m > 10.0, true);
        expect(m > 100.0, false);
        expect(m > 1000.0, false);
        expect(m > double.infinity, false);
        expect(m > double.negativeInfinity, true);
      });

      test('Operator greater or eq than', () {
        final m = Money(100);
        expect(m >= 10.0, true);
        expect(m >= 100.0, true);
        expect(m >= 1000.0, false);
        expect(m > double.infinity, false);
        expect(m > double.negativeInfinity, true);
      });

      test('Operator less than', () {
        final m = Money(100);
        expect(m < 1000.0, true);
        expect(m < 100.0, false);
        expect(m < 10.0, false);
        expect(m < double.infinity, true);
        expect(m < double.negativeInfinity, false);
      });

      test('Operator greater or eq than', () {
        final m = Money(100);
        expect(m <= 10.0, false);
        expect(m <= 100.0, true);
        expect(m <= 1000.0, true);
        expect(m < double.infinity, true);
        expect(m < double.negativeInfinity, false);
      });

      test('Operator eq', () {
        final m = Money(100);
        expect(m == 1000.0, false);
        expect(m == 100.0, true);
        expect(m == 10.0, false);
        expect(m == double.infinity, false);
        expect(m == double.negativeInfinity, false);
        expect(m == double.nan, false);
      });

      test('Operator not eq', () {
        final m = Money(100);
        expect(m != 1000.0, true);
        expect(m != 100.0, false);
        expect(m != 10.0, true);
        expect(m != double.infinity, true);
        expect(m != double.negativeInfinity, true);
        expect(m != double.nan, true);
      });
    });

    group('String', () {
      test('Operator greater than', () {
        final m = Money(100);
        expect(m > '10.0', true);
        expect(m > '100.0', false);
        expect(m > '1000', false);
      });

      test('Operator greater or eq than', () {
        final m = Money(100);
        expect(m >= '10.0', true);
        expect(m >= '100', true);
        expect(m >= '1000.0', false);
      });

      test('Operator less than', () {
        final m = Money(100);
        expect(m < '1000.0', true);
        expect(m < '100.0', false);
        expect(m < '10', false);
      });

      test('Operator less than', () {
        final m = Money(100);
        expect(m <= '1000.0', true);
        expect(m <= '100.0', true);
        expect(m <= '10', false);
      });

      test('Operator eq', () {
        final m = Money(100);
        expect(m == '100.0', true);
        expect(m == '10', false);
      });

      test('Operator not eq', () {
        final m = Money(100);
        expect(m != '1000.0', true);
        expect(m != '100', false);
      });
    });

    group('CompareTo', () {
      group('Int', () {
        test('Int greater argument', () {
          expect(Money.zero.compareTo(1), isNegative);
        });
        test('Int lesser argument', () {
          expect(Money.zero.compareTo(-1), isPositive);
        });
        test('Int eq', () {
          expect(Money.zero.compareTo(0), isZero);
        });
      });

      group('Double', () {
        test('Double greater argument', () {
          expect(Money.zero.compareTo(1.0), isNegative);
        });
        test('Double lesser argument', () {
          expect(Money.zero.compareTo(-1.0), isPositive);
        });
        test('Double eq', () {
          expect(Money.zero.compareTo(0.0), isZero);
        });
        test('Double infinity', () {
          expect(Money.zero.compareTo(double.infinity), isNegative);
        });
        test('Double negative infinity', () {
          expect(Money.zero.compareTo(double.negativeInfinity), isPositive);
        });
        test('Double nan', () {
          expect(() => Money.zero.compareTo(double.nan), throwsA(isA<Error>()));
        });
      });

      group('String', () {
        test('String greater argument', () {
          expect(Money.zero.compareTo('100'), isNegative);
        });
        test('String lesser argument', () {
          expect(Money.zero.compareTo('-100'), isPositive);
        });
        test('String eq', () {
          expect(Money.zero.compareTo('0'), isZero);
        });
      });
    });
  });

  group('Operators', () {
    group('Zero', () {
      test('Add zero', () {
        final z = Money.zero;
        final m = Money(100);
        final result = z + m;
        expect(result.value, 100.0);
        expect(result.raw, 1000000);
      });

      test('Subtract zero', () {
        final z = Money.zero;
        final m = Money(100);
        final result = m - z;
        expect(result.value, 100.0);
        expect(result.raw, 1000000);
      });

      test('Subtract from zero', () {
        final z = Money.zero;
        final m = Money(100);
        final result = z - m;
        expect(result.value, -100.0);
        expect(result.raw, -1000000);
      });

      test('Add', () {
        final result = Money.zero + 1;
        expect(result.value, 1.0);
        expect(result.raw, 10000);
      });

      test('Subtract', () {
        final result = Money.zero - 1;
        expect(result.value, -1.0);
        expect(result.raw, -10000);
      });

      test('Multiply', () {
        final result = Money.zero * 1;
        expect(result.value, isZero);
        expect(result.raw, isZero);
      });

      test('Divide', () {
        final result = Money.zero / 10;
        expect(result.value, isZero);
        expect(result.raw, isZero);
      });
    });

    group('Int', () {
      test('Add positive', () {
        final m = Money(100);
        final result = m + 1;
        expect(result.value, 101.00);
        expect(result.raw, 1010000);
      });

      test('Add negative', () {
        final m = Money(100);
        final result = m + (-1);
        expect(result.value, 99.0);
        expect(result.raw, 990000);
      });

      test('Subtract positive', () {
        final m = Money(100);
        final result = m - 1;
        expect(result.value, 99);
        expect(result.raw, 990000);
      });

      test('Subtract negative', () {
        final m = Money(100);
        final result = m - (-1);
        expect(result.value, 101.00);
        expect(result.raw, 1010000);
      });

      test('Multiply', () {
        final m = Money(100);
        final result = m * 3;
        expect(result.value, 300.0);
        expect(result.raw, 3000000);
      });

      test('Multiply by zero', () {
        final m = Money(100);
        final result = m * 0;
        expect(result.value, isZero);
        expect(result.raw, isZero);
      });

      test('Divide', () {
        final m = Money(100);
        final result = m / 10;
        expect(result.value, 10.0);
        expect(result.raw, 100000);
      });

      test('Divide by zero', () {
        final m = Money(100);
        expect(() => m / 0, throwsA(isA<Error>()));
      });
    });

    group('Double', () {
      test('Add positive', () {
        final m = Money(100);
        final result = m + 1.01;
        expect(result.value, 101.01);
        expect(result.raw, 1010100);
      });

      test('Add negative', () {
        final m = Money(100);
        final result = m + (-1.01);
        expect(result.value, 98.99);
        expect(result.raw, 989900);
      });

      test('Add infinity', () {
        final m = Money(100);
        expect(() => m + double.infinity, throwsA(isA<Error>()));
        expect(() => m + double.negativeInfinity, throwsA(isA<Error>()));
      });

      test('Add nan', () {
        final m = Money(100);
        expect(() => m + double.nan, throwsA(isA<Error>()));
      });

      test('Subtract from zero', () {
        final z = Money.zero;
        final m = Money(100);
        final result = z - m;
        expect(result.value, -100.0);
        expect(result.raw, -1000000);
      });

      test('Subtract positive', () {
        final m = Money(100);
        final result = m - 1.01;
        expect(result.value, 98.99);
        expect(result.raw, 989900);
      });

      test('Subtract negative', () {
        final m = Money(100);
        final result = m - (-1.01);
        expect(result.value, 101.01);
        expect(result.raw, 1010100);
      });

      test('Subtract infinity', () {
        final m = Money(100);
        expect(() => m - double.infinity, throwsA(isA<Error>()));
        expect(() => m - double.negativeInfinity, throwsA(isA<Error>()));
      });

      test('Subtract nan', () {
        final m = Money(100);
        expect(() => m - double.nan, throwsA(isA<Error>()));
      });

      test('Multiply greater than 1 ratio', () {
        final m = Money(100);
        final result = m * 3.1;
        expect(result.value, 310.0);
        expect(result.raw, 3100000);
      });

      test('Multiply less than 1 ratio', () {
        final m = Money(100);
        final result = m * .1;
        expect(result.value, 10.0);
        expect(result.raw, 100000);
      });

      test('Multiply zero ratio', () {
        final m = Money(100);
        final result = m * 0;
        expect(result.value, isZero);
        expect(result.raw, isZero);
      });

      test('Multiply infinity', () {
        final m = Money(100);
        expect(() => m * double.infinity, throwsA(isA<Error>()));
        expect(() => m * double.infinity, throwsA(isA<Error>()));
      });

      test('Multiply nan', () {
        final m = Money(100);
        expect(() => m * double.nan, throwsA(isA<Error>()));
      });

      test('Divide greater than 1 ratio', () {
        final m = Money(100);
        final result = m / 10.0;
        expect(result.value, 10.0);
        expect(result.raw, 100000);
      });

      test('Divide less than 1 ratio', () {
        final m = Money(100);
        final result = m / 0.1;
        expect(result.value, 1000.0);
        expect(result.raw, 10000000);
      });

      test('Divide by zero', () {
        final m = Money(100);
        expect(() => m / 0, throwsA(isA<Error>()));
      });

      test('Divide infinity', () {
        final m = Money(100);
        expect(() => m / double.infinity, throwsA(isA<Error>()));
        expect(() => m / double.negativeInfinity, throwsA(isA<Error>()));
      });

      test('Divide nan', () {
        final m = Money(100);
        expect(() => m / double.nan, throwsA(isA<Error>()));
      });
    });

    group('String', () {
      test('Add positive', () {
        final m = Money(100);
        final result = m + '1.01';
        expect(result.value, 101.01);
        expect(result.raw, 1010100);
      });

      test('Add negative', () {
        final m = Money(100);
        final result = m + '-10';
        expect(result.value, 90.0);
        expect(result.raw, 900000);
      });

      test('Subtract positive', () {
        final m = Money(100);
        final result = m - '19';
        expect(result.value, 81.0);
        expect(result.raw, 810000);
      });

      test('Subtract negative', () {
        final m = Money(100);
        final result = m - '-1.01';
        expect(result.value, 100.99);
        expect(result.raw, 1009900);
      });

      test('Multiply', () {
        final m = Money(100);
        final result = m * '-3.1';
        expect(result.value, -310.0);
        expect(result.raw, -3100000);
      });

      test('Multiply by zero', () {
        final m = Money(100);
        final result = m * '0';
        expect(result.value, isZero);
        expect(result.raw, isZero);
      });

      test('Divide', () {
        final m = Money(100);
        final result = m / 10;
        expect(result.value, 10.0);
        expect(result.raw, 100000);
      });

      test('Divide by zero', () {
        final m = Money(100);
        expect(() => m / '0', throwsA(isA<Error>()));
      });
    });

    test('Unary minus positive', () {
      final m = -Money(1);
      expect(m.value, -1);
    });

    test('Unary minus negative', () {
      final m = -Money(-1);
      expect(m.value, 1);
    });

    test('Unary minus zero', () {
      final m = -Money.zero;
      expect(m.raw, isZero);
    });

    test('Absolute value', () {
      final positive = Money(1).abs();
      final negative = Money(-1).abs();
      expect(positive == negative, isTrue);
    });
  });

  group('Formatter', () {
    test("Change formatter", () {
      final m = Money(-1234.567);
      expect(m.toString(), '-1234.56');
      m.formatter = MoneyFormatter("#`###,@@@s");
      expect(m.toString(), '1`234,567-');
    });
  });
}


test/formatter_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:money/formatter.dart';

void main() {
  group('Parse pattern', () {
    test('Empty pattern', () {
      final f = MoneyFormatter("#");
      expect(() => MoneyFormatter(""), throwsA(isA<Error>()));
    });

    test('Wrong sign symbol', () {
      expect(() => MoneyFormatter("x#"), throwsA(isA<Error>()));
    });

    test('Wrong sign symbols place', () {
      expect(() => MoneyFormatter("sp#"), throwsA(isA<Error>()));
      expect(() => MoneyFormatter("#p"), throwsA(isA<Error>()));
    });

    test('No integral part', () {
      expect(() => MoneyFormatter("s"), throwsA(isA<Error>()));
      expect(() => MoneyFormatter("p"), throwsA(isA<Error>()));
    });

    test('Second sign operator', () {
      expect(() => MoneyFormatter("s#s"), throwsA(isA<Error>()));
      expect(() => MoneyFormatter("p#s"), throwsA(isA<Error>()));
    });

    test('Wrong integral part symbol', () {
      expect(() => MoneyFormatter("%"), throwsA(isA<Error>()));
      expect(() => MoneyFormatter("s%"), throwsA(isA<Error>()));
      expect(() => MoneyFormatter("%.@@"), throwsA(isA<Error>()));
    });

    test('No thousands separator', () {
      expect(() => MoneyFormatter("##"), throwsA(isA<Error>()));
      expect(() => MoneyFormatter("###"), throwsA(isA<Error>()));
      expect(() => MoneyFormatter("####"), throwsA(isA<Error>()));
    });

    test('Wrong thousands separator', () {
      expect(() => MoneyFormatter("#s###"), throwsA(isA<Error>()));
      expect(() => MoneyFormatter("#/###"), throwsA(isA<Error>()));
      expect(() => MoneyFormatter("#''###"), throwsA(isA<Error>()));
      expect(() => MoneyFormatter("#'###'###"), throwsA(isA<Error>()));
      expect(() => MoneyFormatter("#'#"), throwsA(isA<Error>()));
    });

    test('No decimal separator', () {
      expect(() => MoneyFormatter("#@@"), throwsA(isA<Error>()));
    });

    test('Wrong decimal separator', () {
      expect(() => MoneyFormatter("#;@@"), throwsA(isA<Error>()));
      expect(() => MoneyFormatter("#.,@@"), throwsA(isA<Error>()));
    });

    test('Same thousand and decimal separators', () {
      expect(() => MoneyFormatter("#.###.@@"), throwsA(isA<Error>()));
      expect(() => MoneyFormatter("#,###,@@"), throwsA(isA<Error>()));
      expect(MoneyFormatter("#,###.@@"), isNotNull);
    });

    test('No decimal part', () {
      expect(MoneyFormatter("#"), isNotNull);
    });

    test('Sole decimal place', () {
      expect(() => MoneyFormatter("#.@"), throwsA(isA<Error>()));
    });

    test('5 decimal places', () {
      expect(() => MoneyFormatter("#.@@@@@"), throwsA(isA<Error>()));
    });

    test('Wrong decimal place symbol', () {
      expect(() => MoneyFormatter("#.&"), throwsA(isA<Error>()));
    });

    test('Wrong padding operator symbol', () {
      expect(() => MoneyFormatter("#.&@@"), throwsA(isA<Error>()));
      expect(() => MoneyFormatter("#.@@)"), throwsA(isA<Error>()));
      expect(() => MoneyFormatter("#.??@@"), throwsA(isA<Error>()));
      expect(() => MoneyFormatter("#.@@??"), throwsA(isA<Error>()));
    });

    test('Second padding operator', () {
      expect(() => MoneyFormatter("#.?@@?"), throwsA(isA<Error>()));
    });
  });

  test("Negative decimal", () {
    final f = MoneyFormatter("#");
    expect(() => f.format(0, -1), throwsA(isA<Error>()));
  });

  group("Format", () {
    test("Defaults", () {
      final f = MoneyFormatter("#");
      expect(f.format(0, 0), "0.00");
      expect(f.format(10, 1), "10.10");
      expect(f.format(10, 10), "10.10");
      expect(f.format(10, 109), "10.10");
      expect(f.format(1000, 0), "1000.00");
      expect(f.format(-1234, 0), "-1234.00");
    });

    group("Sign", () {
      test("Prefix sign", () {
        final f = MoneyFormatter("s#");
        expect(f.format(-1234, 0), "-1234.00");
      });

      test("Postfix sign", () {
        final f = MoneyFormatter("#s");
        expect(f.format(-1234, 0), "1234.00-");
      });

      test("Parenthesis sign", () {
        final f = MoneyFormatter("p#");
        expect(f.format(-1234, 0), "(1234.00)");
      });
    });

    group("Thousand separator", () {
      test("Thousand separator: - (space) -", () {
        final f = MoneyFormatter("# ###");
        expect(f.format(0, 0), "0.00"); // no separator
        expect(f.format(1000, 0), "1 000.00");
        expect(f.format(-1234, 0), "-1 234.00");
      });

      test("Thousand separator: - ' -", () {
        final f = MoneyFormatter("#'###");
        expect(f.format(999, 0), "999.00"); // no separator
        expect(f.format(1000, 0), "1'000.00");
        expect(f.format(-1234, 0), "-1'234.00");
      });

      test("Thousand separator: - . -", () {
        final f = MoneyFormatter("#.###");
        expect(f.format(999, 0), "999,00"); // no separator
        expect(f.format(1000, 0), "1.000,00");
        expect(f.format(-1234, 0), "-1.234,00");
      });

      test("Thousand separator: - , -", () {
        final f = MoneyFormatter("#,###");
        expect(f.format(999, 0), "999.00"); // no separator
        expect(f.format(1000, 0), "1,000.00");
        expect(f.format(-1234, 0), "-1,234.00");
      });

      test("Thousand separator: - ‘ -", () {
        final f = MoneyFormatter("#‘###");
        expect(f.format(999, 0), "999.00"); // no separator
        expect(f.format(1000, 0), "1‘000.00");
        expect(f.format(-1234, 0), "-1‘234.00");
      });

      test("Thousand separator: - ` -", () {
        final f = MoneyFormatter("#`###");
        expect(f.format(999, 0), "999.00"); // no separator
        expect(f.format(1000, 0), "1`000.00");
        expect(f.format(-1234, 0), "-1`234.00");
      });
    });

    group("Decimals", () {
      test("2 Places", () {
        final f = MoneyFormatter("#.@@");
        expect(f.format(0, 0), "0.00");
        expect(f.format(10, 1), "10.10");
        expect(f.format(10, 101), "10.10");
        expect(f.format(10, 1012), "10.10");
        expect(f.format(10, 1010), "10.10");
        expect(f.format(-1234, 0), "-1234.00");
      });
      test("3 Places", () {
        final f = MoneyFormatter("#.@@@");
        expect(f.format(0, 0), "0.000");
        expect(f.format(10, 1), "10.100");
        expect(f.format(10, 101), "10.101");
        expect(f.format(10, 1012), "10.101");
        expect(f.format(10, 1010), "10.101");
        expect(f.format(-1234, 0), "-1234.000");
      });
      test("4 Places", () {
        final f = MoneyFormatter("#.@@@@");
        expect(f.format(0, 0), "0.0000");
        expect(f.format(10, 1), "10.1000");
        expect(f.format(10, 101), "10.1010");
        expect(f.format(10, 1012), "10.1012");
        expect(f.format(10, 1010), "10.1010");
        expect(f.format(-1234, 0), "-1234.0000");
      });
    });
  });
}


lib/formatter.dart
import 'dart:math';

/// Money formatter
///
/// Numbers, Negative numbers and Thousands separator
/// Decimal places + Padding
///
/// Numbers
///
/// Integral part operator: #
/// Example: # == 123; 0; 1
///
/// Thousand separator operator: #(op)### (4 integral operators and separator)
/// Valid symbols for separator (operator) are
/// period (.), comma (,), space ( ), apostrophe (‘) and single quote (')
/// Example: # ### == 1 234; 0; 123; 1 234 567 890
///          #,### == 1,234; 0; 123; 1,234,567,890
/// Negative numbers are declared as sign or parenthesis
/// Example: p# == (123); (1234)
/// Example: s# == -123 (default)
/// Example: #s == 123- (Netherlands)
/// Example: s#,### == -1,234
/// Example: p#,### == (1,234)
///
/// Decimal places operator @ and decimal part separator (valid symbols are period '.' and comma ',')
/// #.@@, where # - integral part format, '.' - decimal separator, @ - mandatory decimal place (max 4)
/// Example: #.@@ == 123.45; 0.00; -123.00
///
/// Padding operator for decimal places: '?'
/// Padding after decimal separator declare all decimal places are not mandatory
/// and prints all decimal places only if value has decimal part
/// Padding after decimal format prints only significant places
/// Example: #.@@ ==  123.01; 0.00; 0.10 (default)
/// Example: #.?@@ == 123.01; 0; 0.10
/// Example: #.?@@@ == 123.010; 0; 0.100
/// Example: #.@@? == 123.01; 0; 0.1
/// Example: #.@@@? == 123.01; 0; 0.1; 456.789
///
/// Full example: #,###.@@ == 0.00; -1.23; -1,234.56; 1,234,567.80
/// Full example: s#.@@ == 0.00; -1.23; -1234.56; 1234567.80
/// Full example: p#.@@ == 0.00; (1.23); (1234.56); 1234567.80
/// Full example: s#,###.?@@ == 0; -1.23; -1234.56; 1234567.80
/// Full example: s#,###.@@? == 0; -1.23; -1234.56; 1234567.8
///
/// Default sign for decimal separator is period (.) -- todo get it from locale
/// If thousand separator specified as period (.) -- or , from locale
/// and decimal separator not (default use) - formatter will use different decimal separator (, or . accordingly)

class MoneyFormatParseError extends ArgumentError {
  MoneyFormatParseError(String pattern, [String name = "", String message = ""])
      : super.value(pattern, name, message);
}

enum _Sign {
  Sign,
  Parenthesis,
}

enum _Decimals {
  Mandatory,
  NotMandatory,
  MandatorySignificantOnly,
}

class MoneyFormatter {
  static final RegExp _re = RegExp(
      r"(^[s|p]?)((#[\s.,‘`']##)?#)((([.,])([\?]?)(@{2,4})([\?]?))?)([s]?$)");

  final String pattern;

  late final String _decimalSeparator;
  late final String _thousandsSeparator;

  late final _Sign _sign;
  late final bool _signAtEnd;

  late final int _decimalPlaces;
  late final _Decimals _decimalsMandatory;

  MoneyFormatter(this.pattern) {
    _parse();
  }

  // TODO make formatter from locale
  // MoneyFormatter.locale(String locale);

  String format(int integral, int decimal) {
    if (decimal < 0)
      throw ArgumentError.value(decimal, "decimal", "negative decimal");

    bool negative = integral.sign < 0;
    integral = integral.abs();

    String integralPart = "";
    if (integral != 0 && _thousandsSeparator.isNotEmpty) {
      int separator = 3;
      while (integral > 0) {
        integralPart += (integral % 10).toString();
        integral ~/= 10;
        separator -= 1;
        if (separator == 0) {
          integralPart += _thousandsSeparator;
          separator = 3;
        }
      }
      if (integralPart[integralPart.length - 1] == _thousandsSeparator)
        integralPart = integralPart.substring(0, integralPart.length - 1);
      integralPart = integralPart.split('').reversed.join();
    } else {
      integralPart = integral.toString();
    }

    if (decimal > 9999) {
      print("warning: decimal part longer than 4 digits - truncated");
      while (decimal > 9999) decimal ~/= 10;
    }

    final reduceDecimal = (int d) {
      while (d > pow(10, _decimalPlaces) - 1) d ~/= 10; // TODO rounding?
      return d;
    };

    String decimalPart = "";
    if (_decimalsMandatory == _Decimals.Mandatory ||
        (_decimalsMandatory == _Decimals.NotMandatory && decimal != 0)) {
      decimalPart = _decimalSeparator +
          reduceDecimal(decimal).toString().padRight(_decimalPlaces, '0');
    } else if (decimal == 0) {
      // nothing to add - just print integral part
    } else if (_decimalsMandatory == _Decimals.MandatorySignificantOnly) {
      decimalPart = _decimalSeparator + reduceDecimal(decimal).toString();
    } else {
      print("whats here? $integral . $decimal");
    }

    String number = integralPart + decimalPart;

    // sign
    if (negative) {
      if (_sign == _Sign.Parenthesis) {
        number = "($number)";
      } else if (_sign == _Sign.Sign && _signAtEnd) {
        number = "$number-";
      } else {
        number = "-$number";
      }
    }

    return number;
  }

  void _parse() {
    RegExpMatch? match = _re.firstMatch(pattern);
    if (match == null) throw MoneyFormatParseError(pattern, "pattern");

    if (match.groupCount != 10)
      throw MoneyFormatParseError(
          pattern, "pattern", "not enough parsing groups");

    // debug
    // print(pattern);
    // for (var i = 0; i < match.groupCount + 1; i++) {
    //   print("$i: ${match.group(i)}");
    // }

    // Decimal separator
    var decimalSeparator = match.group(6);
    if (decimalSeparator == null || decimalSeparator.isEmpty)
      decimalSeparator = ".,";

    // Decimal places
    final decimals = match.group(8);
    _decimalPlaces =
        (decimals == null || decimals.isEmpty) ? 2 : decimals.length;

    // Decimal padding
    var decimalPartPadding = match.group(7);
    if (decimalPartPadding == null) decimalPartPadding = "";
    var significantOnlyPadding = match.group(9);
    if (significantOnlyPadding == null) significantOnlyPadding = "";
    if (significantOnlyPadding.isNotEmpty && decimalPartPadding.isNotEmpty)
      throw MoneyFormatParseError(
          pattern, "pattern", "error: 2 padding operators '?' in pattern");
    else
      _decimalsMandatory = significantOnlyPadding.isNotEmpty
          ? _Decimals.MandatorySignificantOnly
          : decimalPartPadding.isNotEmpty
              ? _Decimals.NotMandatory
              : _Decimals.Mandatory;

    // Thousands separator
    var thousandSeparator = match.group(3);
    if (thousandSeparator == null) thousandSeparator = "";
    _thousandsSeparator = thousandSeparator.replaceAll("#", "");

    if (decimalSeparator.length > 1) {
      if (_thousandsSeparator.contains(RegExp(r'[,.]')))
        decimalSeparator = decimalSeparator.replaceAll(_thousandsSeparator, "");
      else
        decimalSeparator = "."; // TODO get it from locale
    }

    _decimalSeparator = decimalSeparator;
    if (_thousandsSeparator == _decimalSeparator)
      throw MoneyFormatParseError(pattern, "pattern",
          "thousand '$_thousandsSeparator' and decimal '$_decimalSeparator' separators are equal");

    // Sign
    var prefixSign = match.group(1);
    if (prefixSign == null) prefixSign = "";
    var postfixSign = match.group(10);
    if (postfixSign == null) postfixSign = "";
    if (prefixSign.isEmpty && postfixSign.isEmpty) {
      _sign = _Sign.Sign;
      _signAtEnd = false;
    } else if (prefixSign.isNotEmpty && postfixSign.isNotEmpty) {
      throw MoneyFormatParseError(pattern, "pattern",
          "invalid sign attribute: prefix '$prefixSign' and postfix '$postfixSign'");
    } else if (postfixSign.isNotEmpty) {
      _sign = _Sign.Sign;
      _signAtEnd = true;
    } else {
      _signAtEnd = false;
      _sign = prefixSign == "s" ? _Sign.Sign : _Sign.Parenthesis;
    }
  }
}
